pkg/clientconn/client.go

```diff
@@ -285,7 +285,14 @@ func AuthenticatedGRPCConnection(endpoint string, server mtls.Subject, extraConn
 	}
 
 	var dialOpts []grpc.DialOption
-	dialOpts = append(dialOpts, keepAliveDialOption())
+	// To avoid getting 'Client received GoAway with error code ENHANCE_YOUR_CALM and debug data equal to ASCII "too_many_pings"'
+	// from Sensor in Compliance, we must match the client and server settings for keepalive
+	// See: https://github.com/grpc/grpc/blob/master/doc/keepalive.md#faq
+	dialOpts = append(dialOpts, grpc.WithKeepaliveParams(keepalive.ClientParameters{
+		Time:                10 * time.Second,
+		Timeout:             40 * time.Second,
+		PermitWithoutStream: false,
```

Comment:
Just out of curiosity; why did you change `PermitWithoutStream` to `false`? Couldn't this lead to problems in some scenarios?

Reply:
That was the primary reason for the error. In https://github.com/grpc/grpc/blob/master/doc/keepalive.md#faq, we can read:

> if a server has GRPC_ARG_KEEPALIVE_PERMIT_WITHOUT_CALLS set to false while the client has set this to true resulting in keepalive pings being sent even when there is no call in flight.

Compliance was a client here, so I changed the setting to see if that helps - it does. I think, we have a streams between: (1) Sensor and Central, (2) Collector and Sensor, but not between Compliance and Sensor.

Second factor was:

> if the client's GRPC_ARG_KEEPALIVE_TIME_MS setting is lower than the server's GRPC_ARG_HTTP2_MIN_RECV_PING_INTERVAL_WITHOUT_DATA_MS.

Which I think is the `10s` parameter in that block of code. It was okay, but to be double sure, I changed the timeout from 30s to 40s to match the settings in sensor.

Reply:
Gotcha! That makes sense thank you 🙂